ENH: use non-trivial bounds to solve for kappa of vonmises MLE.

fancidev opened the pull request:
#### Reference issue
This is a follow-up commit for gh-18190.

#### What does this implement/fix?
The MLE of the vonmises distribution solves $r_0(\kappa):=I_1(\kappa)/I_0(\kappa)=r$ for $\kappa$ given $0 < r < 1$ using Brent's method. This PR provides non-trivial bounds for the root.

The non-trivial bounds save a few evaluations of $I_1(\kappa)/I_0(\kappa)$  compared to the trivial bounds. It also opens the possibility to solve the equation to arbitrary precision.

#### Algorithm description

The bounds of $\kappa$ are derived from the bounds of $r_\nu(x):=I_{\nu+1}(x)/I_\nu(x)$ given by [Amos (1973)](https://www.jstor.org/stable/2005830). He gave two set of bounds:

Equation (11):

$$
\frac{x}{\nu+1+\sqrt{x^2+(\nu+1)^2}}\le r_\nu(x)\le\frac{x}{\nu+\sqrt{x^2+(\nu+2)^2}}
$$

Equation (16):

$$
\frac{x}{\nu+0.5+\sqrt{x^2+(\nu+1.5)^2}}\le r_\nu(x)\le\frac{x}{\nu+0.5+\sqrt{x^2+(\nu+0.5)^2}}
$$

The lower bound of (16) is always tighter, while the upper bound of (16) is tighter only if $x\ge 2\sqrt{(\nu+1)(2\nu+3)}$.

From these bounds we derive the bounds for the solution $x=r_\nu^{-1}(r)$:

Equation (11’):

$$
\frac{r\nu+r\sqrt{\nu^2+4(\nu+1)(1-r^2)}}{1-r^2}\le r_\nu^{-1}(r)\le \frac{r(2\nu+2)}{1-r^2}
$$

Equation (16’):

$$
\frac{r(2\nu+1)}{1-r^2}\le r_\nu^{-1}(r) \le \frac{r(\nu+0.5)+r\sqrt{(\nu+0.5)^2+2(\nu+1)(1-r^2)}}{1-r^2}
$$

The upper bound of (16’) is always tighter, while the lower bound of (16’) is tighter only if $r\ge \frac{1}{2}\sqrt{(2\nu+3)/(\nu+1)}$.

The bounds are moderate for moderate values of $r$, but remarkably tight for small and large $r$. For example, as $r$ tends to 0, the ratio between the upper bound of (16’) and the lower bound of (11’) tends to 1. And as $r$ tends to 1, the width of the bounds (16’) tends to $(2\nu+2)/(2\nu+1)$.

When $\nu=0$, the bounds become

Equation (11’’):

$$
\frac{2r}{\sqrt{1-r^2}}\le r_0^{-1}(r) \le \frac{2r}{1-r^2}
$$

Equation (16’’):

$$
\frac{r}{1-r^2} \le r_0^{-1}(r) \le \frac{\frac{1}{2}r+r\sqrt{\frac{1}{4}+2(1-r^2)}}{1-r^2}
$$

Any (relaxation) of these  bounds may be used to bracket the root for Brent’s method. A convenient choice is the lower bound of (16’’) together with the upper bound of (11’’); this leads to

$$
\frac{r}{1-r^2} \le r_0^{-1}(r) \le \frac{2r}{1-r^2}
$$

This choice trivially guarantees that the lower bound is less than the upper bound numerically.

#### Additional information

[Segura (2023)](https://www.sciencedirect.com/science/article/pii/S0022247X23002147) gives an in-depth analysis of the bounds of $r_\nu(x)$.

[Hill (1981)](https://dl.acm.org/doi/10.1145/355945.355949) gives series expansions for $r_0(x)$, and uses Newton-Raphson to invert it.

fancidev committed:
ENH: use non-trivial bounds to solve for kappa of vonmises MLE.

dschmitz89 commented on scipy/stats/_continuous_distns.py:
```diff
@@ -10664,12 +10664,32 @@ def find_kappa(data, loc):
+                # [1] Amos, D. E. (1973).  Computation of Modified Bessel
+                #     Functions and Their Ratios.  Mathematics of Computation,
+                #     28(125): 239-251.
+                with np.errstate(divide='ignore'):
```
When does the error occur? The following snippet executes without a warning for me:

```python
import numpy as np

r = 0.999999999999999
2*r/np.sqrt(1-r*r), 2*r/(1-r*r)
``` 

Side note: one thing I read in some other PR here a while back was that writing $a^2-b^2=(a+b)(a-b)$ typically yields better precision with floating point, so we could do `(1+r)(1-r)` instead of `1-r*r`. Not sure how relevant that is here though, feel free to ignore.

fancidev replied:
It sounds on the safe side to use `(1+r)(1-r)`, and I'll use that.

And you raised a very good point about "stress testing" the root finder. It does reveal the current bounds are too "tight" numerically. Half the lower bound and double the upper bound appears to pass my stress tests. I'll put the test code in a separate comment for easy reference.

dschmitz89 reviewed (commented):
Thanks @fancidev , looks very good.

I have one inline question about the errorstate. To be on the safe side, could you run locally tests such as [here](https://github.com/scipy/scipy/pull/18824)? I am pretty sure, this update is good but we have been bitten too often in the past with seemingly easy changes.

dschmitz89 commented:
The same bounds could be applied to the Von Mises-Fisher distribution: https://github.com/scipy/scipy/blob/6103deb58885562adcaee50307e0cef006d84a75/scipy/stats/_multivariate.py#L6871

fancidev commented:
The non-trivial bounds are tested numerically for different values of $r$ between $0$ and $1$, with more values near 0 and 1. The experiments reveal that the mathematical bounds are too "tight" numerically -- the function value at the bounds have the same sign for nearly half of the test cases! The previous trivial bounds (`np.finfo(float).tiny`) also fails 2% of the test cases.

The numerical issues indicate that the precision of `i1e(x)/i0e(x)` is not very good. For now the issues can be worked around in one of two ways. One is to half the lower bound and double the upper bound. The other is to set the lower bound to 0 and upper bound to 1e16.

```python
import numpy as np
import scipy.stats
import scipy.special

rs = np.unique(np.concatenate([
    np.linspace(0, 1, 10000),
    np.geomspace(np.nextafter(0, 1), np.nextafter(1, 0), 10000),
    1 - np.geomspace(np.nextafter(0, 1), np.nextafter(1, 0), 10000),
]))
errors = []
count = 0
for i, r in enumerate(rs):
    def solve_for_kappa(kappa):
        return scipy.special.i1e(kappa)/scipy.special.i0e(kappa)-r
    
    MAX_KAPPA = 1e16
    # MAX_KAPPA = np.inf

    with np.errstate(divide='ignore'):
        # lower_bound = min(MAX_KAPPA, 2*r/np.sqrt((1-r)*(1+r)) / 2)
        # upper_bound = min(MAX_KAPPA, 2*r/((1-r)*(1+r)) * 2)
        lower_bound = min(MAX_KAPPA, 2*r/np.sqrt((1-r)*(1+r)) / 2)
        upper_bound = min(MAX_KAPPA, 2*r/((1-r)*(1+r)) * 2)
    # lower_bound = np.finfo(float).tiny
    # lower_bound = 0
    # upper_bound = 1e16
    try:
        root = scipy.optimize.root_scalar(
            solve_for_kappa, method="brentq", bracket=(lower_bound, upper_bound)).root
    except:
        print(f'{i=} {r=} f({lower_bound})={solve_for_kappa(lower_bound)} '
              f'f({upper_bound})={solve_for_kappa(upper_bound)}')
        count += 1
    assert np.isfinite(root)
    errors.append(solve_for_kappa(root))

print(f'# failures: {count} / {len(rs)}')
print(f'Max abs error: {np.max(np.abs(errors))}')
```

Output:

```
# failures: 0 / 20449
Max abs error: 3.229287411218372e-13
```
```

fancidev committed:
ENH: improve numerical accuracy of the bounds of kappa.

- A simpler lower bound is used, which is in addition tigher for large r.
  The bound is derived from the upper bound of Eq. 16 in Amos (1973).

- Numerical violation of the bounds are handled.  This happens for
  certain values of small and large r, where I[1](x)/I[0](x) is very
  flat and the numerical evaluation loses precision.  The violated
  bound is returned in these cases as they are more accurate.

- Branch an explicit case for r == 1, and remove unnecessary checks
  for the case 0 < r < 1 to improve readability.

dschmitz89 commented:
Thanks for the thorough analysis @fancidev ! This is a big improvement over what we had before. 2 points left:

- Could you add tests for the new cases (`r=1` and data for which the bounds are the solution already)?
- Could you add Amos` paper to the references section of the von mises docs?

fancidev commented:
Sure

> Thanks for the thorough analysis @fancidev ! This is a big improvement over what we had before. 2 points left:
> 
> * Could you add tests for the new cases (`r=1` and data for which the bounds are the solution already)?
> * Could you add Amos` paper to the references section of the von mises docs?

Sure, let me add test case(s) to cover the branches.

Regarding the docs, are you referring to the [reference page](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.vonmises.html) or the [tutorial page](https://docs.scipy.org/doc/scipy/tutorial/stats/continuous_vonmises.html)? (The tutorial may be a more suitable place, but doesn’t often appear first in google, and there’s no link from the reference page to the tutorial page. Ideally both pages could be merged and enriched.)

dschmitz89 commented:
> > Thanks for the thorough analysis @fancidev ! This is a big improvement over what we had before. 2 points left:
> > 
> > * Could you add tests for the new cases (`r=1` and data for which the bounds are the solution already)?
> > * Could you add Amos` paper to the references section of the von mises docs?
> 
> Sure, let me add test case(s) to cover the branches.
> 
> Regarding the docs, are you referring to the [reference page](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.vonmises.html) or the [tutorial page](https://docs.scipy.org/doc/scipy/tutorial/stats/continuous_vonmises.html)? (The tutorial may be a more suitable place, but doesn’t often appear first in google, and there’s no link from the reference page to the tutorial page. Ideally both pages could be merged and enriched.)

I meant the reference page: see [here](https://github.com/scipy/scipy/blob/4edfcaa3ce8a387450b6efce968572def71be089/scipy/stats/_continuous_distns.py#L1669) for an example (sphinx can have a confusing syntax if you are not already familiar with it).

fancidev committed:
TST: add test case for vonmises mle where all data are equal.

fancidev committed:
TST: test that root bracket violation is handled in vonmises mle.

fancidev commented:
@dschmitz89 I added test cases for the new branches.

As for adding “Amos (1973)” to the reference page, I feel it doesn’t fit the topic there as the paper is about a technical detail in inverting the kappa MLE equation. Describing the MLE algorithm is out of the scope of the reference page. An interested user could browse the source code, where I have put the reference in. 

Should we ever add a special function for inverting Bessel I function ratio, the Amos paper would appear in the reference page there :-) How do you think?

dschmitz89 reviewed (changes requested):
I noticed some issues still when comparing against the generic fit method, will update later. Cannot find how to dismiss my old review (Github ..).

fancidev commented:
> I noticed some issues still when comparing against the generic fit method, will update later. Cannot find how to dismiss my old review (Github ..).

Thanks for checking! Maybe you can just click “Resolve conversation” in the old reviews.

In addition, I see there’re some merge conflicts now. Shall I make a new, rebased commit to make things easier?

dschmitz89 commented:
Sorry for the late reply @fancidev , I am having trouble with my local setup at the moment.

I performed a test against the generic (`superfit`) method and found a significant number of cases where this PR results in slightly worse log likelihood:

```python
import numpy as np
from scipy import stats
from time import perf_counter
import matplotlib.pyplot as plt

rng = np.random.default_rng(2433006235492611347)
qrng = stats.qmc.Halton(4, seed=rng)

N = 1000
times_pr = np.full(N, np.nan)
times_super = np.full(N, np.nan)
nllfs_super = np.full(N, np.nan)
nllfs_pr = np.full(N, np.nan)

for i in range(1000):
    print(i)
    sample = qrng.random()
    sample = stats.qmc.scale(sample, [-np.pi, -np.pi, -10, 1], [np.pi, np.pi, 15, 5])[0]
    #print(sample)

    loc, floc, shape, size = sample
    shape = 10**shape
    size = int(10**size)
    #print(f"loc: {loc}, kappa: {shape}")
    
    dist = stats.vonmises(loc=loc, kappa=shape)
    rvs = dist.rvs(size=size, random_state=rng)

    try:
        tic = perf_counter()
        shape_fit, loc_fit, dummy = stats.vonmises.fit(rvs)#, floc=floc)
        #print(f"loc fit: {loc_fit}")
        #print(f"shape fit: {shape_fit}")
        toc = perf_counter()
        nllf_pr = stats.vonmises.nnlf((shape_fit, loc_fit, dummy), rvs)
        #print(f"Override. loc fit: {loc_fit}, shape fit: {shape_fit}, logpdf: {nllf_pr}")
        #print(f"nnlf: {nllf_pr}")
        nllfs_pr[i] = nllf_pr
        times_pr[i] = toc-tic

    except Exception as e:
        print(f"Failure: {e}")

    try:
        tic = perf_counter()
        shape_fit, loc_fit, dummy = stats.vonmises.fit(rvs, superfit=True, fscale=1.0)#, floc=floc)
        #print(f"loc fit super: {loc_fit}")
        #print(f"shape fit super: {shape_fit}")
        toc = perf_counter()
        nllf_super = stats.vonmises.nnlf((shape_fit, loc_fit, dummy), rvs)
        #print(f"Super. loc fit: {loc_fit}, shape fit: {shape_fit}, logpdf: {nllf_super}")
        nllfs_super[i] = nllf_super
        times_super[i] = toc-tic

    except Exception as e:
        print(f"Failure: {e}")

nllf_diff = (nllfs_pr - nllfs_super)/np.abs(nllfs_super)
nllf_diff = nllf_diff[np.isfinite(nllf_diff)]
nllf_diff_good = -nllf_diff[nllf_diff < 0]
nllf_diff_bad = nllf_diff[nllf_diff > 0]
print(nllf_diff_bad)
bins = np.arange(-6, 1.5, 0.25)
plt.hist(np.log10(times_pr), bins=bins, alpha=0.5)
plt.hist(np.log10(times_super), bins=bins, alpha=0.5)
plt.legend(('PR', 'main'))
plt.xlabel('log10 of `fit` execution time (s)')
plt.ylabel('Frequency')
plt.title('Histogram of log10 of `fit` execution time (s)') #, loc fixed
plt.savefig("./VonMises_Time.png", dpi=300)

plt.clf()
plt.hist(np.log10(nllf_diff_good), label='Good', alpha=0.5)
plt.hist(np.log10(nllf_diff_bad), label='Bad', alpha=0.5)
plt.xlabel('log10 of NLLF changes')
plt.ylabel('Frequency')
plt.title('Histogram of log10 of NLLF relative improvement') #, loc fixed
plt.legend()
plt.savefig("VonMises_NLLF_Differences.png", dpi=300)
``` 

![VonMises_NLLF_Differences_PR](https://github.com/scipy/scipy/assets/40656107/86d0cd50-2739-4f36-ba2c-0a94fc2564ec)

I did not dig deeper due to issues with my setup, so I do not know in which regime for $\kappa$ the issues occur. Could you check if expanding the root finding bracket solves the issue?

fancidev commented:
Thanks for the detailed comparison @dschmitz89 !

The difference in fitted log likelihood value between `fit()` and `fit(superfit=True)` shows a similar histogram (like the one you posted above) using the production Python code, so it's not introduced by this PR. Yet it's interesting to investigate them further.

Those relative errors of the order `1e-16` to `1e-14` are not important as they are on the order of numerical noise. But some larger ones are intriguing. The largest relative error (about 6%) is reproduced by the following code:

```python
import numpy as np
from scipy import stats
x = np.array([-1.32782601, -1.32782601, -1.32782604, -1.32782607, -1.3278261 ,
       -1.32782606, -1.32782605, -1.32782611, -1.32782613, -1.3278261 ,
       -1.32782608, -1.32782608, -1.32782603, -1.32782608, -1.32782608,
       -1.32782613, -1.32782604, -1.3278261 , -1.32782611, -1.32782608,
       -1.32782601, -1.3278261 , -1.32782607, -1.32782604, -1.32782605,
       -1.3278261 , -1.32782607, -1.32782606, -1.32782605, -1.3278261 ,
       -1.327826  , -1.32782603, -1.32782611, -1.32782601, -1.32782603,
       -1.32782603, -1.32782606, -1.32782609, -1.32782608, -1.32782602,
       -1.32782608, -1.32782597, -1.32782607, -1.32782608, -1.32782604,
       -1.32782607, -1.32782608, -1.32782607, -1.32782606, -1.32782603,
       -1.32782612, -1.32782603, -1.32782607, -1.32782603, -1.32782615,
       -1.32782603, -1.32782609, -1.32782611, -1.327826  , -1.32782611,
       -1.327826  , -1.32782605, -1.32782603, -1.32782603, -1.32782606,
       -1.32782603, -1.3278261 , -1.32782601, -1.32782602, -1.32782601,
       -1.32782606, -1.32782598, -1.32782605, -1.32782606, -1.32782607,
       -1.32782605, -1.32782602, -1.327826  , -1.3278261 , -1.32782606,
       -1.32782602, -1.32782608, -1.32782605, -1.32782612, -1.32782607,
       -1.32782606, -1.32782602, -1.32782615, -1.32782608, -1.32782606,
       -1.32782603, -1.32782602, -1.32782607, -1.32782606, -1.32782606,
       -1.32782607, -1.32782604, -1.32782604, -1.32782601, -1.32782605,
       -1.32782602, -1.32782604])
args0 = stats.vonmises.fit(x, fscale=1, superfit=True)
args1 = stats.vonmises.fit(x)
print(f'Super fit: {args0[0:2]}, nllf={stats.vonmises.nnlf(args0, x)}')
print(f'Child fit: {args1[0:2]}, nllf={stats.vonmises.nnlf(args1, x)}')
```

Output:

```
Super fit: (45049734747248.06, -1.3278260801942854), nllf=-1505.4054408910633
Child fit: (976562499999999.5, -1.3278260578431365), nllf=-1599.4828608677947
```

So this is the region of `kappa` where the function `I[1](x)/I[0](x)` becomes very flat, and it appears some numerical issues exist. Let me take a closer look to fix it!

dschmitz89 commented:
@fancidev : interesting, indeed a numerical problem. If issues only occur for extreme values for `kappa`, we might as well leave it as it is before we over optimize this, what do you think? Do issues also occur for `kappa<1e8` for example?

dschmitz89 commented:
One reason could be that the Bessel ratio is not monotonous for high values, will check in the coming days. Such issues we saw with other special functions too.

dschmitz89 commented:
Ok, I am pretty sure now that the issue is the precision of the Bessel ratio for very high concentration parameters:

```python
import matplotlib.pyplot as plt
import numpy as np
import scipy
x = np.logspace(13, 14, 1000)
y = scipy.special.i1e(x)/scipy.special.i0e(x)
plt.semilogx(x, y)
plt.show()
``` 

Zooming in, I get the following plot:

![kappa](https://github.com/scipy/scipy/assets/40656107/854feeb1-e57c-42cd-94af-be90d0a46ebd)

@fancidev : I think the issues for very high values can be ignored as implementing a special function for that purpose would be a big undertaking. For lower to moderate values, your PR is a big performance improvement. Let's resolve the merge conflicts and then merge this into main?

fancidev commented:
Thanks for the analysis @dschmitz89 ! I rebased the changes and they should be able to merge now.

I'll look into the remaining numerical noise issue afterwards.

Thanks again for reviewing!

dschmitz89 commented:
Thanks again @fancidev ! All your recent contributions are highly appreciated. 🚀 

Can I interest you in doing a similar PR for the Von Mises-Fisher distrubution [xref](https://github.com/scipy/scipy/pull/20102#issuecomment-1951059513)? Otherwise, I will look into it in the coming weeks.

dschmitz89 commented:
There were still some linting issues actually such as missing whitespace around arithmetic operators but these issues seem to be completely neglected by ruff since it expects the code to get autoformatted. Insisting on correcting code style that is not caught by the linter will potentially annoy many contributors, so I am probably not going to be very persistent with such things in reviews. Other opinions appreciated though.

fancidev commented:
> Can I interest you in doing a similar PR for the Von Mises-Fisher distrubution [xref](https://github.com/scipy/scipy/pull/20102#issuecomment-1951059513)? Otherwise, I will look into it in the coming weeks.

Sure. I’ll look into those after spending a few hours to see if the numerical noise with vonmises MLE can be mitigated properly. In fact I took a quick look at `vonmises_fisher` before; it requires much more polishing work than vonmises. To begin with, `ive(kappa)` returns `nan` soon enough before `kappa` gets large enough to be out of the domain of concern.

As for linting, I just make sure my PyCharm doesn’t draw squiggly lines :) Spacing around arithmetic operators is not required by PEP8, and PEP8 doesn’t like systematic spacing around binary operators such as `c = (a + b) * (a - b)`.
